### PR TITLE
Fix headless login with `second_factor: on | optional`

### PR DESCRIPTION
**Pull request by Joerger**
Update headless login to work with `second_factor: on | optional`, rather than only `second_factor: webauthn`. Headless login should work with any standard webauthn configuration, but I missed this configuration in the original implementation.

### CODE REVIEW

**Review by mdwn — APPROVED** [review]

File: lib/web/apiserver.go [hunk 1]
```diff
@@ -3122,9 +3122,21 @@ func (h *Handler) createSSHCert(w http.ResponseWriter, r *http.Request, p httpro
...
-		PublicKey:      req.PubKey,
-		ClientMetadata: clientMetaFromReq(r),
+		HeadlessAuthenticationID: req.HeadlessAuthenticationID,
+		Username:                 req.User,
+		PublicKey:                req.PubKey,
+		ClientMetadata:           clientMetaFromReq(r),
+	}
+
+	if authReq.HeadlessAuthenticationID != "" {
+		// create a new http client with a standard callback timeout.
```
Comment (codingllama) [inline lib/web/apiserver.go:3132]: Why? What's wrong with using the existing client? Mind explaining that in the comment as well?

File: lib/web/apiserver.go [hunk 2]
```diff
@@ -3138,17 +3150,7 @@ func (h *Handler) createSSHCert(w http.ResponseWriter, r *http.Request, p httpro
...
-
-		// create a new http client with a standard callback timeout.
-		authenticationClient, err = authClient.CloneHTTPClient(
-			auth.ClientParamTimeout(defaults.CallbackTimeout),
-			auth.ClientParamResponseHeaderTimeout(defaults.CallbackTimeout),
-		)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
+		// WebAuthn supports this endpoint for headless login.
```
Comment (codingllama) [inline lib/web/apiserver.go:3153]: I'm not sure I follow this comment. Rephrase?

Shouldn't this block initialize some WebAuthn-specific structure?

**Review by jakule — APPROVED** [review]

File: lib/web/apiserver.go (hunk 1 above)
Comment (Joerger) [reply lib/web/apiserver.go]: Updated the comment

File: lib/web/apiserver.go (hunk 2 above)
Comment (Joerger) [reply lib/web/apiserver.go]: Removed in favor of handling headless authentication as a special case above.

File: lib/auth/methods.go [hunk 3]
```diff
@@ -202,6 +202,18 @@ func (s *Server) authenticateUser(ctx context.Context, req AuthenticateUserReque
 	var authErr error // error message kept obscure on purpose, use logging for details
 	switch {
 	// cases in order of preference
+	case req.HeadlessAuthenticationID != "":
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/methods.go:205]: Added empty Webauthn and OTP fields to `TestServer_Authenticate_headless` to test that headless login is prefered.

[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/methods.go:205]: I'll take that, but I would prefer testing the non-locking aspects of it as well.

**Review by codingllama — APPROVED** [review]

**Review by codingllama — APPROVED** [review]

File: lib/auth/methods.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/methods.go:205]: Thanks!

**Pull request merged**